File: python3/Binary Search/weighted_random_selection.py
import random
from typing import List


class WeightedRandomSelection:
    def __init__(self, weights: List[int]):
        self.prefix_sums = [weights[0]]
        for i in range(1, len(weights)):
            self.prefix_sums.append(self.prefix_sums[-1] + weights[i])

    def select(self) -> int:
        # Pick a random target between 1 and the largest endpoint on the number 
        # line.
        target = random.randint(1, self.prefix_sums[-1])
        left, right = 0, len(self.prefix_sums) - 1
        # Perform lower-bound binary search to find which endpoint (i.e., prefix 
        # sum value) corresponds to the target.
        while left < right:
            mid = (left + right) // 2
            if self.prefix_sums[mid] < target:
                left = mid + 1
            else:
                right = mid
        return left

File: csharp/Binary Search/WeightedRandomSelection.cs
public class WeightedRandomSelection
{
    private readonly int[] _prefixSums;
    private readonly Random _random;

    public WeightedRandomSelection(int[] weights)
    {
        _prefixSums = new int[weights.Length];
        _prefixSums[0] = weights[0];

        for (int i = 1; i < weights.Length; i++)
            _prefixSums[i] = _prefixSums[i - 1] + weights[i];

        _random = new Random();
    }

    public int Select()
    {
        // Pick a random target between 1 and the largest endpoint on the number line.
        int target = _random.Next(1, _prefixSums[^1]);
        int left = 0, right = _prefixSums.Length - 1;

        // Perform lower-bound binary search to find which endpoint (i.e., prefix 
        // sum value) corresponds to the target.
        while (left < right)
        {
            int mid = (left + right) / 2;

            if (_prefixSums[mid] < target)
                left = mid + 1;
            else
                right = mid;
        }

        return left;
    }
}
